### PR TITLE
shares.py: don't use recursion when scanning shares

### LINKED ISSUE #2560: RecursionError: maximum recursion depth exceeded while calling a Python object when trying to share files
Nicotine+ version: 3.2.9 
Operating System/Distribution:
Debian 12, Flatpak

When I try to share files, I get the following message:
```
2023-06-29 20:32:08 Serious error occurred while rescanning shares. If this problem persists, delete /home/moon/.var/app/org.nicotine_plus.Nicotine/data/nicotine/*.db and try again. If that doesn't help, please file a bug report with this stack trace included: 
Traceback (most recent call last):
  File "/app/lib/python3.10/site-packages/pynicotine/shares.py", line 113, in run
    new_mtimes, new_files, new_streams = self.rescan_dirs("normal", rebuild=self.rebuild)
  File "/app/lib/python3.10/site-packages/pynicotine/shares.py", line 249, in rescan_dirs
    files, streams, mtimes = self.get_files_list(
  File "/app/lib/python3.10/site-packages/pynicotine/shares.py", line 366, in get_files_list
    dir_files, dir_streams, dir_mtimes = self.get_files_list(
  File "/app/lib/python3.10/site-packages/pynicotine/shares.py", line 366, in get_files_list
    dir_files, dir_streams, dir_mtimes = self.get_files_list(
  File "/app/lib/python3.10/site-packages/pynicotine/shares.py", line 366, in get_files_list
    dir_files, dir_streams, dir_mtimes = self.get_files_list(
  [Previous line repeated 986 more times]
  File "/app/lib/python3.10/site-packages/pynicotine/shares.py", line 363, in get_files_list
    if self.is_hidden(path, entry=entry):
  File "/app/lib/python3.10/site-packages/pynicotine/shares.py", line 286, in is_hidden
    last_folder = os.path.basename(os.path.normpath(folder.replace('\\', '/')))
  File "/usr/lib/python3.10/posixpath.py", line 143, in basename
    sep = _get_sep(p)
  File "/usr/lib/python3.10/posixpath.py", line 42, in _get_sep
    if isinstance(path, bytes):
RecursionError: maximum recursion depth exceeded while calling a Python object
```
I tried giving it all files permissions and linked the external drive folder, but nothing has changed.

### PR DESCRIPTION
**Pull request by mathiascode**
Fixes #2560

### CODE REVIEW

File: pynicotine/shares.py [hunk 1]
```diff
@@ -367,75 +368,73 @@ def is_hidden(folder, filename=None, entry=None):
...
+        mtimes = {}
+        folders = deque([(shared_folder, None)])
 
-        if not rebuild and folder in oldmtimes and mtime == oldmtimes[folder]:
-            try:
-                files[virtual_folder] = oldfiles[virtual_folder]
-                streams[virtual_folder] = oldstreams[virtual_folder]
-                folder_unchanged = True
+        while folders:
+            folder, folder_stat = folders.pop()
```
Comment (slook) [inline pynicotine/shares.py:380]: Is it intentional that the while loop exits before the last item in the deque?

Comment (mathiascode) [reply pynicotine/shares.py:380]: Are you sure? That shouldn't be the case.

Comment (slook) [reply pynicotine/shares.py:380]: It only from reviewing the code, that it appears like the code after this statement will be unreachable for the final item.

However, no share items appear to be missed. I must be misunderstanding.

Comment (mathiascode) [reply pynicotine/shares.py:380]: while loops don't terminate immediately when the condition becomes false, they terminate when `while folders` is evaluated again.

**Pull request merged**